scipy/spatial/ckdtree/src/partial_sort.h

```diff
@@ -1,29 +1,35 @@
-
 #ifndef CKDTREE_PARTIAL_SORT
 #define CKDTREE_PARTIAL_SORT
 
-/* Splitting routines for a balanced kd-tree
- * Code originally written by Jake Vanderplas for scikit-learn
+/* Adapted version of the code originally 
+ * written by @jiefangxuanyan for scikit-learn.
  *
  */
 
-inline void
-index_swap(ckdtree_intp_t *arr, intptr_t i1, intptr_t i2)
-{
-    /* swap the values at index i1 and i2 of arr */
-    ckdtree_intp_t tmp = arr[i1];
-    arr[i1] = arr[i2];
-    arr[i2] = tmp;
-}
 
-static void
-partition_node_indices(const double *data,
-                       ckdtree_intp_t *node_indices,
-                       ckdtree_intp_t split_dim,
-                       ckdtree_intp_t split_index,
-                       ckdtree_intp_t n_features,
-                       ckdtree_intp_t n_points)
-{
+#include <algorithm>
+
+class IndexComparator {
+    const double *data;
+    ckdtree_intp_t split_dim, n_features;
+
+public:
+    IndexComparator(const double *data, ckdtree_intp_t split_dim, ckdtree_intp_t n_features):
+        data(data), split_dim(split_dim), n_features(n_features) {}
```

Comment:
Apparently split_dim takes value in range [0, n_features). If so, should we prefer n_dims to n_features, or add a comment?

Reply:
@rainwoodman, is more suitable here, I know. However, in the original version of the function declaration, it was n_features, and I decided to leave it. I investigated the code, and it seems that nothing prevents to switch to n_dims. In the newest version I used "n_dims" instead of "n_features".